code/datums/components/manual_heart.dm

```diff
@@ -0,0 +1,170 @@
+	UnregisterSignal(parent, list(COMSIG_CARBON_GAIN_ORGAN, COMSIG_CARBON_LOSE_ORGAN, COMSIG_LIVING_REVIVE, COMSIG_LIVING_DEATH, SIGNAL_ADDTRAIT(TRAIT_NOBLOOD), SIGNAL_REMOVETRAIT(TRAIT_NOBLOOD)))
+
+/datum/component/manual_heart/proc/restart()
+	SIGNAL_HANDLER
+
+	if(check_valid())
+		last_pump = world.time
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS) //make sure the action button always shows as available when it is
+		START_PROCESSING(SSdcs, src)
+
+/datum/component/manual_heart/proc/pause()
+	SIGNAL_HANDLER
+	pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+	STOP_PROCESSING(SSdcs, src)
+
+/// Worker proc that checks logic for if a pump can happen, and applies effects/notifications from doing so
+/datum/component/manual_heart/proc/on_pump(mob/owner)
+	last_pump = world.time
+	playsound(owner,'sound/effects/singlebeat.ogg', 40, TRUE)
+	owner.balloon_alert(owner, "your heart beats")
+
+	if(!iscarbon(owner))
+		return
+	var/mob/living/carbon/carbon_owner = owner
+
+	if(HAS_TRAIT(carbon_owner, TRAIT_NOBLOOD))
+		return
+	carbon_owner.blood_volume = min(carbon_owner.blood_volume + (blood_loss * 0.5), BLOOD_VOLUME_MAXIMUM)
+	carbon_owner.remove_client_colour(/datum/client_colour/manual_heart_blood)
+	add_colour = TRUE
+	carbon_owner.adjustBruteLoss(-heal_brute)
+	carbon_owner.adjustFireLoss(-heal_burn)
+	carbon_owner.adjustOxyLoss(-heal_oxy)
+
+/datum/component/manual_heart/process()
+	var/mob/living/carbon/carbon_parent = parent
+
+	//If they aren't connected, don't kill them.
+	if(!istype(carbon_parent) || !carbon_parent.client)
+		last_pump = world.time
+		return
+
+	if(world.time <= (last_pump + pump_delay))
+		return
+
+	carbon_parent.blood_volume = max(carbon_parent.blood_volume - blood_loss, 0)
+	to_chat(carbon_parent, span_userdanger("You have to keep pumping your blood!"))
+	last_pump = world.time - (2 SECONDS) //give two full seconds before losing more blood
+	if(add_colour)
+		carbon_parent.add_client_colour(/datum/client_colour/manual_heart_blood)
+		add_colour = FALSE
+
+///If a new heart is added, start processing.
+/datum/component/manual_heart/proc/check_added_organ(mob/organ_owner, obj/item/organ/new_organ)
+	SIGNAL_HANDLER
+
+	var/obj/item/organ/internal/heart/new_heart = new_organ
+
+	if(istype(new_heart) && check_valid())
+		last_pump = world.time
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+		START_PROCESSING(SSdcs, src)
+
+///If the heart is removed, stop processing.
+/datum/component/manual_heart/proc/check_removed_organ(mob/organ_owner, obj/item/organ/removed_organ)
+	SIGNAL_HANDLER
+
+	var/obj/item/organ/internal/heart/removed_heart = removed_organ
+
+	if(istype(removed_heart))
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+		STOP_PROCESSING(SSdcs, src)
+
+///Helper proc to check if processing can be restarted.
+/datum/component/manual_heart/proc/check_valid()
+	var/mob/living/carbon/carbon_parent = parent
+	if(!istype(parent))
```

Comment:
I've asked this in a couple places but are you _sure_ we need to keep checking that we didn't stop being a carbon?